### PR TITLE
PLFM-4120

### PR DESCRIPTION
**Pull request by brucehoff**

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java [hunk 1]
```diff
@@ -136,11 +138,17 @@ public DockerAuthorizationToken authorizeDockerAccess(UserInfo userInfo, String
...
+					String parentId = validParentProjectId(repositoryPath);
+					if (parentId==null) {
+						// can't push to a non-existent parent
+						as = AuthorizationManagerUtil.ACCESS_DENIED;
+					} else {
+						// check for create permission on parent
+						Node node = new Node();
+						node.setParentId(parentId);
+						node.setNodeType(EntityType.dockerrepo);
+						as = authorizationManager.canCreate(userInfo, node);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java:150]: refactor to canCreate to take the two params that are actually used.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java [hunk 2]
```diff
@@ -152,11 +160,15 @@ public DockerAuthorizationToken authorizeDockerAccess(UserInfo userInfo, String
 				}
 				break;
 			case pull:
-				// check DOWNLOAD permission and add to permittedActions
-				if (existingDockerRepoId!=null) {
-					if (authorizationManager.canAccess(
-							userInfo, existingDockerRepoId, ObjectType.ENTITY, ACCESS_TYPE.DOWNLOAD).getAuthorized()
-					) permittedActions.add(requestedAction);
+				if (
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java:163]: might need a test for repository in the trash.

File: lib/jdomodels/src/main/java/org/sagebionetworks/evaluation/dbo/SubmissionDBO.java [hunk 3]
```diff
@@ -294,6 +321,19 @@ public MigrationType getMigratableTableType() {
 			@Override
 			public SubmissionDBO createDatabaseObjectFromBackup(
 					SubmissionDBO backup) {
+				if (backup.getDockerRepositoryName()==null) {
+					try {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/evaluation/dbo/SubmissionDBO.java:325]: need jira to remove this after migration

File: lib/jdomodels/src/main/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImpl.java [hunk 4]
```diff
@@ -228,6 +243,20 @@
 
 	private static final String SUBSTATUS_IN_CLAUSE = " AND ss."+COL_SUBSTATUS_STATUS+" IN (:"+COL_SUBSTATUS_STATUS+")";
 
+	private static final String SUBMISSIONS_WITH_DOCKER_REPO_AND_PERMISSION_SQL =
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImpl.java:246]: I prefer the join syntax.

File: lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImplTest.java [hunk 5]
```diff
@@ -189,12 +211,24 @@ public void setUp() throws DatastoreException, InvalidModelException, NotFoundEx
...
+        acl.setCreatedBy(userId);
+        acl.setCreationDate(new Date());
+        acl.setId(evalId);
+        Set<ResourceAccess> ras = new HashSet<ResourceAccess>();
+        ResourceAccess ra = new ResourceAccess();
+        ra.setPrincipalId(Long.parseLong(userId));
+        ra.setAccessType(ModelConstants.EVALUATION_ADMIN_ACCESS_PERMISSIONS);
+        ras.add(ra);
+        acl.setResourceAccess(ras);
+        acl.setId(aclDAO.create(acl, ObjectType.EVALUATION));
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImplTest.java:224]: consider using a Util for the creation of the ACL.

File: lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImplTest.java [hunk 6]
```diff
@@ -239,13 +273,22 @@ public void tearDown() throws DatastoreException, NotFoundException  {
...
+		try {
 			evaluationDAO.delete(evalId);
 		} catch (NotFoundException e) {};
 		
     	try {
     		nodeDAO.delete(nodeId);
     	} catch (NotFoundException e) {};
-		fileHandleDAO.delete(fileHandleId);
+    	
+		// TEMPORARY if (fileHandleId!=null) fileHandleDAO.delete(fileHandleId);
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImplTest.java:287]: should be live code

File: lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dbo/SubmissionDBOTest.java [hunk 7]
```diff
@@ -135,5 +141,23 @@ public void testCRD() throws Exception{
...
+    	DockerRepository entity = new DockerRepository();
+    	entity.setRepositoryName(dockerRepositoryName);
+    	bundle.setEntity(entity);
+    	JSONObjectAdapter joa = new JSONObjectAdapterImpl();
+    	bundle.writeToJSONObject(joa);
+
+    	backup.setEntityBundle(joa.toJSONString().getBytes());
+    	SubmissionDBO dbo = backup.getTranslator().createDatabaseObjectFromBackup(backup);
+    	assertEquals(dockerRepositoryName, dbo.getDockerRepositoryName());
+    }
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dbo/SubmissionDBOTest.java:162]: test for case where data is already migrated.

Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dbo/SubmissionDBOTest.java:162]: test where object is not a docker repo

File: services/repository-managers/src/test/java/org/sagebionetworks/evaluation/manager/SubmissionManagerTest.java [hunk 8]
```diff
@@ -488,26 +492,47 @@ public void testCreateWithErroneousDockerDigest() throws Exception{
...
+		JSONObjectAdapter joa = new JSONObjectAdapterImpl();
+		bundle.writeToJSONObject(joa);
+		return joa.toJSONString();
+	}
+	
 	@Test
 	public void testCreateDockerRepoSubmission() throws Exception {
 		sub.setEntityId(DOCKER_REPO_ENTITY_ID);
+		EntityBundle dockerBundle = createDockerEntityBundle();
+		sub.setEntityBundleJSON(createEntityBundleJSON(dockerBundle));
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/evaluation/manager/SubmissionManagerTest.java:514]: these two lines can be in before()

File: lib/jdomodels/src/main/java/org/sagebionetworks/evaluation/dbo/SubmissionDBO.java (hunk 3 above)
Comment (brucehoff) [reply lib/jdomodels/src/main/java/org/sagebionetworks/evaluation/dbo/SubmissionDBO.java:325]: https://sagebionetworks.jira.com/browse/PLFM-4363

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java (hunk 1 above)
Comment (brucehoff) [reply services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java:150]: Done.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java (hunk 2 above)
Comment (brucehoff) [reply services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java:163]: We reviewed the tests and concluded that this case is covered.

File: lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImplTest.java (hunk 5 above)
Comment (brucehoff) [reply lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImplTest.java:224]: DONE

File: lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImplTest.java (hunk 6 above)
Comment (brucehoff) [reply lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImplTest.java:287]: DONE

File: lib/jdomodels/src/main/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImpl.java (hunk 4 above)
Comment (brucehoff) [reply lib/jdomodels/src/main/java/org/sagebionetworks/evaluation/dao/SubmissionDAOImpl.java:246]: DONE

File: lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dbo/SubmissionDBOTest.java (hunk 7 above)
Comment (brucehoff) [reply lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dbo/SubmissionDBOTest.java:162]: Done.

Comment (brucehoff) [reply lib/jdomodels/src/test/java/org/sagebionetworks/evaluation/dbo/SubmissionDBOTest.java:162]: Done.

**Pull request merged**